packages/synapse-client/src/codegen/ShortenCanonicalName.ts

```diff
@@ -0,0 +1,123 @@
+import { countBy, isEqual, upperFirst } from 'lodash-es'
+
+/**
+ * Given a set of "shortened" canonical names, returns just those that are duplicated.
+ *
+ * Example input: [
+ *  ['foo', 'bar'],
+ *  ['foo', 'baz'],
+ *  ['foo', 'bar']
+ * ]
+ *
+ * Output: [['foo', 'bar']]
+ *
+ * @param canonicalNames
+ */
+function getDuplicateNames(canonicalNames: string[][]): string[][] {
+  // countBy returns an object with the count of each unique value
+  const counts = countBy(
+    Object.values(canonicalNames).map(name =>
+      // re-join the name parts so lodash can compare strings
+      name.join('.'),
+    ),
+  )
+  return (
+    Object.entries(counts)
+      // Return just those that have a count greater than 1
+      .filter(([k, v]) => v > 1)
+      // And convert the name strings back to arrays
+      .map(([k, v]) => k.split('.'))
+  )
+}
+
+/**
+ * Shortens the canonical names in the mapping to the shortest unique name.
+ * Each key in the mapping will be returned in the new object to have a value that is the shortest unique, case-insensitive name for that key.
+ *
+ * Example input: `['org.sagebionetworks.foo.Bar', 'org.sagebionetworks.foo.Baz', 'org.sagebionetworks.bar.Bar']`
+ * Output:
+ * ```
+ *  {
+ *    'org.sagebionetworks.foo.Bar': 'FooBar',
+ *    'org.sagebionetworks.foo.Baz': 'Baz',
+ *    'org.sagebionetworks.bar.Bar': 'BarBar',
+ *  }
+ * ```
+ */
+export function shortenCanonicalNames(
```

Comment:
This function is used to generate the names of the models so we don't have to use model names like `OrgSagebionetworksRepoModelFileEntity`